### ISSUE TITLE
Understanding how the loss is computed

### ISSUE THREAD

**Issue by dariocazzani** [NVIDIA/flowtron#64]
Hi @rafaelvalle 
I was trying to understand how the loss is computed.

I always first try to come out with ideas just by reading the paper, then look into the code and see if my understanding was correct.

I would have expected a sort `MSE` or `MAE` between the output of the network and the original mel-spectrograms, together with a penalty for the `z` output for being far from a spherical gaussian distribution.

In the code I see this:
```
loss = torch.sum(z*z)/(2*self.sigma*self.sigma) - log_s_total
```

What I don't understand is how the network is learning to reproduce the origin mel input, since the loss never directly sees it.

Thank you in advance for your help.

**Comment by rafaelvalle** [comment]
We're learning an **invertible** map from data distribution to a known distribution (Gaussian). Since the map is invertible (bijection), by learning a map from data distribution to a Gaussian distribution, we also learn a map from Gaussian distribution to the data distribution.

**Comment by ciscodario** [comment]
How is the output `z` guaranteed to be gaussian?

**Comment by rafaelvalle** [comment]
It is not guaranteed. During training we're fitting a standard Gaussian N(0, 1) given data, conditions and model parameters.
The better the model is in terms of likelihood, the more z = model(mel, text, sid) will be zero mean and 1 std.

**Comment by ciscodario** [comment]
Exactly, but I still don't understand how that loss is helping with that

**Comment by rafaelvalle** [comment]
With z ~ N(0, 1), maximizing the likelihood of this Gaussian is equivalent to minimizing the length of z times a constant.

The second term, `log_s_total`, comes from the change of variables from one distribution to another.

**Comment by ciscodario** [comment]
I think I get it now, thanks!

**Comment by dariocazzani** [comment]
I think I get it now, thanks! (and sorry for having used 2 different accounts)
